File: python/mirascope/llm/providers/openai/provider.py
"""Unified OpenAI client implementation."""

from collections.abc import Sequence
from typing_extensions import Unpack

from openai import OpenAI

from ...context import Context, DepsT
from ...formatting import Format, FormattableT
from ...messages import Message
from ...responses import (
    AsyncContextResponse,
    AsyncContextStreamResponse,
    AsyncResponse,
    AsyncStreamResponse,
    ContextResponse,
    ContextStreamResponse,
    Response,
    StreamResponse,
)
from ...tools import (
    AsyncContextTool,
    AsyncContextToolkit,
    AsyncTool,
    AsyncToolkit,
    ContextTool,
    ContextToolkit,
    Tool,
    Toolkit,
)
from ..base import BaseProvider, Params
from .completions import OpenAICompletionsProvider
from .model_id import OPENAI_KNOWN_MODELS, OpenAIModelId
from .responses import OpenAIResponsesProvider


def _has_audio_content(messages: Sequence[Message]) -> bool:
    """Returns whether a sequence of messages contains any audio content."""
    for message in messages:
        if message.role == "system":
            continue
        for content in message.content:
            if content.type == "audio":
                return True
    return False


def choose_api_mode(model_id: OpenAIModelId, messages: Sequence[Message]) -> str:
    """Choose between 'responses' or 'completions' API based on model_id and messages.

    Args:
        model_id: The model identifier.
        messages: The messages to send to the LLM.

    Returns:
        Either "responses" or "completions" depending on the model and message content.

    If the user manually specified an api mode (by appending it as a suffix to the model
    id), then we use it.

    Otherwise, we prefer the responses API where supported (because it has better
    reasoning support and better prompt caching). However we will use the :completions api
    if the messages contain any audio content, as audio content is not yet supported in
    the responses API.
    """
    if model_id.endswith(":completions"):
        return "completions"
    elif model_id.endswith(":responses"):
        return "responses"

    if _has_audio_content(messages):
        return "completions"

    if f"{model_id}:responses" in OPENAI_KNOWN_MODELS:
        # Prefer responses api when we know it is available
        return "responses"
    elif f"{model_id}:completions" in OPENAI_KNOWN_MODELS:
        # If we know from testing that the completions api is available, and
        # (implied by above) that responses wasn't, then we should use completions
        return "completions"

    # If we don't have either :responses or :completions in the known_models, it's
    # likely that this is a new model we haven't tested. We default to responses api for
    # openai/ models (on the assumption that they are new models and OpenAI prefers
    # the responses API) but completions for other models (on the assumption that they
    # are other models routing through the OpenAI completions API)
    if model_id.startswith("openai/"):
        return "responses"
    else:
        return "completions"


class OpenAIRoutedCompletionsProvider(OpenAICompletionsProvider):
    """OpenAI completions client that reports provider_id as 'openai'."""

    id = "openai"


class OpenAIRoutedResponsesProvider(OpenAIResponsesProvider):
    """OpenAI responses client that reports provider_id as 'openai'."""

    id = "openai"


class OpenAIProvider(BaseProvider[OpenAI]):
    """Unified provider for OpenAI that routes to Completions or Responses API based on model_id."""

    id = "openai"
    default_scope = "openai/"

    def __init__(
        self, *, api_key: str | None = None, base_url: str | None = None
    ) -> None:
        """Initialize the OpenAI provider with both subclients."""
        self._completions_provider = OpenAIRoutedCompletionsProvider(
            api_key=api_key, base_url=base_url
        )
        self._responses_provider = OpenAIRoutedResponsesProvider(
            api_key=api_key, base_url=base_url
        )
        # Use completions client's underlying OpenAI client as the main one
        self.client = self._completions_provider.client

    def _choose_subprovider(
        self, model_id: OpenAIModelId, messages: Sequence[Message]
    ) -> OpenAICompletionsProvider | OpenAIResponsesProvider:
        """Choose the appropriate provider based on model_id and messages.

        Args:
            model_id: The model identifier.
            messages: The messages to send to the LLM.

        Returns:
            The responses or completions subclient.
        """
        api_mode = choose_api_mode(model_id, messages)
        if api_mode == "responses":
            return self._responses_provider
        return self._completions_provider

    def _call(
        self,
        *,
        model_id: OpenAIModelId,
        messages: Sequence[Message],
        tools: Sequence[Tool] | Toolkit | None = None,
        format: type[FormattableT] | Format[FormattableT] | None = None,
        **params: Unpack[Params],
    ) -> Response | Response[FormattableT]:
        """Generate an `llm.Response` by synchronously calling the OpenAI API.

        Args:
            model_id: Model identifier to use.
            messages: Messages to send to the LLM.
            tools: Optional tools that the model may invoke.
            format: Optional response format specifier.
            **params: Additional parameters to configure output (e.g. temperature). See `llm.Params`.

        Returns:
            An `llm.Response` object containing the LLM-generated content.
        """
        client = self._choose_subprovider(model_id, messages)
        return client.call(
            model_id=model_id,
            messages=messages,
            tools=tools,
            format=format,
            **params,
        )

    def _context_call(
        self,
        *,
        ctx: Context[DepsT],
        model_id: OpenAIModelId,
        messages: Sequence[Message],
        tools: Sequence[Tool | ContextTool[DepsT]]
        | ContextToolkit[DepsT]
        | None = None,
        format: type[FormattableT] | Format[FormattableT] | None = None,
        **params: Unpack[Params],
    ) -> ContextResponse[DepsT, None] | ContextResponse[DepsT, FormattableT]:
        """Generate an `llm.ContextResponse` by synchronously calling the OpenAI API.

        Args:
            ctx: Context object with dependencies for tools.
            model_id: Model identifier to use.
            messages: Messages to send to the LLM.
            tools: Optional tools that the model may invoke.
            format: Optional response format specifier.
            **params: Additional parameters to configure output (e.g. temperature). See `llm.Params`.

        Returns:
            An `llm.ContextResponse` object containing the LLM-generated content.
        """
        client = self._choose_subprovider(model_id, messages)
        return client.context_call(
            ctx=ctx,
            model_id=model_id,
            messages=messages,
            tools=tools,
            format=format,
            **params,
        )

    async def _call_async(
        self,
        *,
        model_id: OpenAIModelId,
        messages: Sequence[Message],
        tools: Sequence[AsyncTool] | AsyncToolkit | None = None,
        format: type[FormattableT] | Format[FormattableT] | None = None,
        **params: Unpack[Params],
    ) -> AsyncResponse | AsyncResponse[FormattableT]:
        """Generate an `llm.AsyncResponse` by asynchronously calling the OpenAI API.

        Args:
            model_id: Model identifier to use.
            messages: Messages to send to the LLM.
            tools: Optional tools that the model may invoke.
            format: Optional response format specifier.
            **params: Additional parameters to configure output (e.g. temperature). See `llm.Params`.

        Returns:
            An `llm.AsyncResponse` object containing the LLM-generated content.
        """
        return await self._choose_subprovider(model_id, messages).call_async(
            model_id=model_id,
            messages=messages,
            tools=tools,
            format=format,
            **params,
        )

    async def _context_call_async(
        self,
        *,
        ctx: Context[DepsT],
        model_id: OpenAIModelId,
        messages: Sequence[Message],
        tools: Sequence[AsyncTool | AsyncContextTool[DepsT]]
        | AsyncContextToolkit[DepsT]
        | None = None,
        format: type[FormattableT] | Format[FormattableT] | None = None,
        **params: Unpack[Params],
    ) -> AsyncContextResponse[DepsT, None] | AsyncContextResponse[DepsT, FormattableT]:
        """Generate an `llm.AsyncContextResponse` by asynchronously calling the OpenAI API.

        Args:
            ctx: Context object with dependencies for tools.
            model_id: Model identifier to use.
            messages: Messages to send to the LLM.
            tools: Optional tools that the model may invoke.
            format: Optional response format specifier.
            **params: Additional parameters to configure output (e.g. temperature). See `llm.Params`.

        Returns:
            An `llm.AsyncContextResponse` object containing the LLM-generated content.
        """
        return await self._choose_subprovider(model_id, messages).context_call_async(
            ctx=ctx,
            model_id=model_id,
            messages=messages,
            tools=tools,
            format=format,
            **params,
        )

    def _stream(
        self,
        *,
        model_id: OpenAIModelId,
        messages: Sequence[Message],
        tools: Sequence[Tool] | Toolkit | None = None,
        format: type[FormattableT] | Format[FormattableT] | None = None,
        **params: Unpack[Params],
    ) -> StreamResponse | StreamResponse[FormattableT]:
        """Generate an `llm.StreamResponse` by synchronously streaming from the OpenAI API.

        Args:
            model_id: Model identifier to use.
            messages: Messages to send to the LLM.
            tools: Optional tools that the model may invoke.
            format: Optional response format specifier.
            **params: Additional parameters to configure output (e.g. temperature). See `llm.Params`.

        Returns:
            An `llm.StreamResponse` object for iterating over the LLM-generated content.
        """
        client = self._choose_subprovider(model_id, messages)
        return client.stream(
            model_id=model_id,
            messages=messages,
            tools=tools,
            format=format,
            **params,
        )

    def _context_stream(
        self,
        *,
        ctx: Context[DepsT],
        model_id: OpenAIModelId,
        messages: Sequence[Message],
        tools: Sequence[Tool | ContextTool[DepsT]]
        | ContextToolkit[DepsT]
        | None = None,
        format: type[FormattableT] | Format[FormattableT] | None = None,
        **params: Unpack[Params],
    ) -> ContextStreamResponse[DepsT] | ContextStreamResponse[DepsT, FormattableT]:
        """Generate an `llm.ContextStreamResponse` by synchronously streaming from the OpenAI API.

        Args:
            ctx: Context object with dependencies for tools.
            model_id: Model identifier to use.
            messages: Messages to send to the LLM.
            tools: Optional tools that the model may invoke.
            format: Optional response format specifier.
            **params: Additional parameters to configure output (e.g. temperature). See `llm.Params`.

        Returns:
            An `llm.ContextStreamResponse` object for iterating over the LLM-generated content.
        """
        client = self._choose_subprovider(model_id, messages)
        return client.context_stream(
            ctx=ctx,
            model_id=model_id,
            messages=messages,
            tools=tools,
            format=format,
            **params,
        )

    async def _stream_async(
        self,
        *,
        model_id: OpenAIModelId,
        messages: Sequence[Message],
        tools: Sequence[AsyncTool] | AsyncToolkit | None = None,
        format: type[FormattableT] | Format[FormattableT] | None = None,
        **params: Unpack[Params],
    ) -> AsyncStreamResponse | AsyncStreamResponse[FormattableT]:
        """Generate an `llm.AsyncStreamResponse` by asynchronously streaming from the OpenAI API.

        Args:
            model_id: Model identifier to use.
            messages: Messages to send to the LLM.
            tools: Optional tools that the model may invoke.
            format: Optional response format specifier.
            **params: Additional parameters to configure output (e.g. temperature). See `llm.Params`.

        Returns:
            An `llm.AsyncStreamResponse` object for asynchronously iterating over the LLM-generated content.
        """
        return await self._choose_subprovider(model_id, messages).stream_async(
            model_id=model_id,
            messages=messages,
            tools=tools,
            format=format,
            **params,
        )

    async def _context_stream_async(
        self,
        *,
        ctx: Context[DepsT],
        model_id: OpenAIModelId,
        messages: Sequence[Message],
        tools: Sequence[AsyncTool | AsyncContextTool[DepsT]]
        | AsyncContextToolkit[DepsT]
        | None = None,
        format: type[FormattableT] | Format[FormattableT] | None = None,
        **params: Unpack[Params],
    ) -> (
        AsyncContextStreamResponse[DepsT]
        | AsyncContextStreamResponse[DepsT, FormattableT]
    ):
        """Generate an `llm.AsyncContextStreamResponse` by asynchronously streaming from the OpenAI API.

        Args:
            ctx: Context object with dependencies for tools.
            model_id: Model identifier to use.
            messages: Messages to send to the LLM.
            tools: Optional tools that the model may invoke.
            format: Optional response format specifier.
            **params: Additional parameters to configure output (e.g. temperature). See `llm.Params`.

        Returns:
            An `llm.AsyncContextStreamResponse` object for asynchronously iterating over the LLM-generated content.
        """
        return await self._choose_subprovider(model_id, messages).context_stream_async(
            ctx=ctx,
            model_id=model_id,
            messages=messages,
            tools=tools,
            format=format,
            **params,
        )

File: typescript/src/llm/providers/openai/provider.ts
/**
 * OpenAI provider implementation with routing between Completions and Responses APIs.
 */

import OpenAI from 'openai';

import type { Context } from '@/llm/context';
import type { Message } from '@/llm/messages';
import type { Params } from '@/llm/models';
import { BaseProvider } from '@/llm/providers/base';
import type { Tools, ContextTools } from '@/llm/tools';
import { OPENAI_ERROR_MAP } from '@/llm/providers/openai/_utils/errors';
import { OpenAICompletionsProvider } from '@/llm/providers/openai/completions/provider';
import { OpenAIResponsesProvider } from '@/llm/providers/openai/responses/provider';
import type { ApiMode, OpenAIModelId } from '@/llm/providers/openai/model-id';
import { OPENAI_KNOWN_MODELS } from '@/llm/providers/openai/model-info';
import { Response } from '@/llm/responses';
import type { ContextResponse } from '@/llm/responses/context-response';
import type { ContextStreamResponse } from '@/llm/responses/context-stream-response';
import { StreamResponse } from '@/llm/responses/stream-response';

/**
 * Check if messages contain any audio content.
 *
 * Audio content requires the Completions API since Responses API doesn't support it yet.
 */
function hasAudioContent(messages: readonly Message[]): boolean {
  for (const message of messages) {
    if (message.role === 'system') {
      continue;
    }
    for (const part of message.content) {
      if (part.type === 'audio') {
        return true;
      }
    }
  }
  return false;
}

/**
 * Choose between 'responses' or 'completions' API based on model ID and messages.
 *
 * If the user manually specified an API mode (by appending it as a suffix to the model
 * ID), then we use it.
 *
 * Otherwise, we prefer the Responses API where supported (because it has better
 * reasoning support and better prompt caching). However we will use the Completions API
 * if the messages contain any audio content, as audio content is not yet supported in
 * the Responses API.
 *
 * @param modelId - The model identifier
 * @param messages - The messages to send to the LLM
 * @returns Either "responses" or "completions" depending on the model and message content
 */
export function chooseApiMode(
  modelId: OpenAIModelId,
  messages: readonly Message[]
): ApiMode {
  // Check explicit suffix
  if (modelId.endsWith(':completions')) {
    return 'completions';
  }
  if (modelId.endsWith(':responses')) {
    return 'responses';
  }

  // Audio content requires completions API
  if (hasAudioContent(messages)) {
    return 'completions';
  }

  // Prefer responses API when we know it is available
  if (OPENAI_KNOWN_MODELS.has(`${modelId}:responses`)) {
    return 'responses';
  }

  // If we know from testing that the completions API is available, and
  // (implied by above) that responses wasn't, then we should use completions
  if (OPENAI_KNOWN_MODELS.has(`${modelId}:completions`)) {
    return 'completions';
  }

  // If we don't have either :responses or :completions in the known models, it's
  // likely that this is a new model we haven't tested. We default to responses API for
  // openai/ models (on the assumption that they are new models and OpenAI prefers
  // the responses API) but completions for other models (on the assumption that they
  // are other models routing through the OpenAI completions API)
  if (modelId.startsWith('openai/')) {
    return 'responses';
  }

  return 'completions';
}

/**
 * Provider for the OpenAI API with intelligent routing between Completions and Responses APIs.
 *
 * This provider automatically selects the appropriate API based on the model ID and messages:
 * - Model IDs ending with `:completions` force the Chat Completions API
 * - Model IDs ending with `:responses` force the Responses API
 * - Messages with audio content use the Completions API (Responses doesn't support audio)
 * - Known models prefer the Responses API when available
 * - Unknown openai/ models default to Responses API
 * - Other unknown models default to Completions API (for OpenAI-compatible providers)
 *
 * @example
 * ```typescript
 * const provider = new OpenAIProvider();
 *
 * // Smart routing: uses Responses API for known models
 * const response = await provider.call({
 *   modelId: 'openai/gpt-4o',
 *   messages: [user('Hello!')],
 * });
 *
 * // Explicitly force Completions API
 * const response2 = await provider.call({
 *   modelId: 'openai/gpt-4o:completions',
 *   messages: [user('Hello!')],
 * });
 * ```
 */
export class OpenAIProvider extends BaseProvider {
  readonly id = 'openai' as const;
  protected readonly errorMap = OPENAI_ERROR_MAP;

  private readonly completionsProvider: OpenAICompletionsProvider;
  private readonly responsesProvider: OpenAIResponsesProvider;

  /**
   * Create a new OpenAI provider instance.
   *
   * @param init - Configuration options
   * @param init.apiKey - OpenAI API key (defaults to OPENAI_API_KEY env var)
   * @param init.baseURL - Optional custom base URL for the API
   */
  constructor(init: { apiKey?: string; baseURL?: string } = {}) {
    super();
    this.completionsProvider = new OpenAICompletionsProvider(init);
    this.responsesProvider = new OpenAIResponsesProvider(init);
  }

  /**
   * Execute a call to the OpenAI API, routing to the appropriate API based on model ID.
   *
   * @param args - Call arguments
   * @param args.modelId - The OpenAI model ID to use (optionally with :completions or :responses suffix)
   * @param args.messages - Array of messages to send
   * @param args.tools - Optional tools to make available to the model
   * @param args.params - Optional additional parameters
   * @returns Response object containing the API response
   */
  protected async _call(args: {
    modelId: string;
    messages: readonly Message[];
    tools?: Tools;
    params?: Params;
  }): Promise<Response> {
    const modelId = args.modelId as OpenAIModelId;
    const apiMode = chooseApiMode(modelId, args.messages);

    if (apiMode === 'responses') {
      // Delegate to the responses provider
      return this.responsesProvider.call({
        modelId: args.modelId,
        messages: args.messages,
        tools: args.tools,
        params: args.params,
      });
    }

    // Delegate to the completions provider
    return this.completionsProvider.call({
      modelId: args.modelId,
      messages: args.messages,
      tools: args.tools,
      params: args.params,
    });
  }

  /**
   * Execute a streaming call to the OpenAI API, routing to the appropriate API based on model ID.
   *
   * @param args - Call arguments
   * @param args.modelId - The OpenAI model ID to use (optionally with :completions or :responses suffix)
   * @param args.messages - Array of messages to send
   * @param args.tools - Optional tools to make available to the model
   * @param args.params - Optional additional parameters
   * @returns StreamResponse object for streaming consumption
   */
  protected async _stream(args: {
    modelId: string;
    messages: readonly Message[];
    tools?: Tools;
    params?: Params;
  }): Promise<StreamResponse> {
    const modelId = args.modelId as OpenAIModelId;
    const apiMode = chooseApiMode(modelId, args.messages);

    if (apiMode === 'responses') {
      return this.responsesProvider.stream({
        modelId: args.modelId,
        messages: args.messages,
        tools: args.tools,
        params: args.params,
      });
    }

    return this.completionsProvider.stream({
      modelId: args.modelId,
      messages: args.messages,
      tools: args.tools,
      params: args.params,
    });
  }

  /**
   * Execute a context-aware call to the OpenAI API, routing to the appropriate API.
   *
   * NOTE: This implementation intentionally duplicates _call() routing logic rather
   * than delegating. When context-aware tools are implemented, this method will
   * diverge to handle passing context to tools during execution. We keep them
   * separate now to make that future change clearer.
   *
   * @param args - Call arguments including context and model
   * @param args.modelId - The OpenAI model ID to use (optionally with :completions or :responses suffix)
   * @param args.messages - Array of messages to send
   * @param args.tools - Optional tools to make available to the model
   * @param args.params - Optional additional parameters
   * @returns ContextResponse object containing the API response
   */
  protected async _contextCall<DepsT>(args: {
    ctx: Context<DepsT>;
    modelId: string;
    messages: readonly Message[];
    tools?: ContextTools<DepsT>;
    params?: Params;
  }): Promise<ContextResponse<DepsT>> {
    const modelId = args.modelId as OpenAIModelId;
    const apiMode = chooseApiMode(modelId, args.messages);

    if (apiMode === 'responses') {
      // Delegate to the responses provider's context call
      return this.responsesProvider.contextCall({
        ctx: args.ctx,
        modelId: args.modelId,
        messages: args.messages,
        tools: args.tools,
        params: args.params,
      });
    }

    // Delegate to the completions provider's context call
    return this.completionsProvider.contextCall({
      ctx: args.ctx,
      modelId: args.modelId,
      messages: args.messages,
      tools: args.tools,
      params: args.params,
    });
  }

  /**
   * Execute a context-aware streaming call to the OpenAI API, routing to the appropriate API.
   *
   * NOTE: This implementation intentionally duplicates _stream() routing logic rather
   * than delegating. When context-aware tools are implemented, this method will
   * diverge to handle passing context to tools during execution. We keep them
   * separate now to make that future change clearer.
   *
   * @param args - Call arguments including context and model
   * @param args.modelId - The OpenAI model ID to use (optionally with :completions or :responses suffix)
   * @param args.messages - Array of messages to send
   * @param args.tools - Optional tools to make available to the model
   * @param args.params - Optional additional parameters
   * @returns ContextStreamResponse object for streaming consumption
   */
  protected async _contextStream<DepsT>(args: {
    ctx: Context<DepsT>;
    modelId: string;
    messages: readonly Message[];
    tools?: ContextTools<DepsT>;
    params?: Params;
  }): Promise<ContextStreamResponse<DepsT>> {
    const modelId = args.modelId as OpenAIModelId;
    const apiMode = chooseApiMode(modelId, args.messages);

    if (apiMode === 'responses') {
      return this.responsesProvider.contextStream({
        ctx: args.ctx,
        modelId: args.modelId,
        messages: args.messages,
        tools: args.tools,
        params: args.params,
      });
    }

    return this.completionsProvider.contextStream({
      ctx: args.ctx,
      modelId: args.modelId,
      messages: args.messages,
      tools: args.tools,
      params: args.params,
    });
  }

  /**
   * Extract the HTTP status code from an OpenAI API error.
   *
   * @param e - The error to extract the status from
   * @returns The HTTP status code or undefined if not available
   */
  /* v8 ignore start - error handling delegated to completions provider */
  protected getErrorStatus(e: Error): number | undefined {
    // OpenAI SDK uses 'status' property
    return (e as InstanceType<typeof OpenAI.APIError>).status;
  }
  /* v8 ignore stop */
}
